python/lsst/ts/vent/controller/controller.py

```diff
@@ -0,0 +1,314 @@
+                slave=self.vfd_slave, address=vfd.Registers.LFR_REGISTER
+            )
+        ).registers[0]
+        return 0.1 * lfr
+
+    async def set_fan_frequency(self, frequency: float) -> None:
+        """Sets the target frequency for the dome exhaust fan. The frequency
+        must be between zero and VFD_MAX_FREQ.
+
+        Parameters
+        ==========
+        frequency the desired VFD frequency in Hz
+
+        Raises
+        ======
+        ValueError if the frequency is not between zero and VFD_MAX_FREQ
+
+        ModbusException if a communications error occurs
+        """
+        logger.debug("set fan_frequency")
+        if not 0 <= frequency <= cfg.VFD_MAX_FREQ:
+            raise ValueError(f"Frequency must be between 0 and {cfg.VFD_MAX_FREQ}")
+
+        settings = {
+            vfd.Registers.CMD_REGISTER: 0 if frequency == 0.0 else 1,
+            vfd.Registers.LFR_REGISTER: round(frequency * 10),
+        }
+        for address, value in settings.items():
+            await self.vfd_client.write_register(slave=self.vfd_slave, address=address, value=value)
+
+    async def vfd_fault_reset(self) -> None:
+        """Resets a fault condition on the VFD so that it will operate again.
+
+        Raises
+        ======
+        ModbusException if a communications error occurs
+        """
+
+        for address, value in vfd.FAULT_RESET_SEQUENCE:
+            await self.vfd_client.write_register(slave=self.vfd_slave, address=address, value=value)
+
+    async def last8faults(self) -> list[tuple[int, str]]:
+        """Returns the last eight fault conditions recorded by the VFD
+
+        Returns
+        =======
+        A list containing 8 tuples each with length 2. The first element in the
+        tuple is an integer fault code, and the second is a human-readable
+        description of the fault code.
+
+        Raises
+        ======
+        ModbusException if a communications error occurs
+        """
+
+        logger.debug("last8faults")
+        rvals = await self.vfd_client.read_holding_registers(
+            slave=1, address=vfd.Registers.FAULT_REGISTER, count=8
+        )
+        return [(r, vfd.VFD_FAULTS[r]) for r in rvals.registers]
+
+    def vent_open(self, vent_number: int) -> None:
+        """Opens the specified vent.
+
+        Parameters
+        ==========
+        vent_number the choice of vent to open, from 0 to 3
+
+        Raises
+        ======
+        ValueError if vent_number is invalid
+
+        RuntimeError if a communications error occurs
+        """
+
+        logger.debug(f"vent_open({vent_number})")
+        if not 0 <= vent_number <= 3:
+            raise ValueError(f"Invalid {vent_number=} should be between 0 and 3")
+        if cfg.VENT_SIGNAL_CH[vent_number] == -1:
+            raise ValueError(f"Vent {vent_number=} is not configured.")
+        megaind.set0_10Out(cfg.MEGAIND_STACK, cfg.VENT_SIGNAL_CH[vent_number], 10.0)
```

Comment:
These are not the correct hardware outputs -- the actuator relays are wired to the open-drain digital outputs (`megaind.setOd`)